---
title: Exploring the causal inference of shear stress associated DNA methylation on cardiovascular risk
author: '**Ruben Methorst, Bsc**, prof. dr. Gerard Pasterkamp, dr. Sander W. Van der Laan'
date: "10-02-2020"
output:
  html_document:
    code_folding: hide
    collapse: yes
    theme: lumen
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
  pdf_document:
    toc: yes
---
## Introduction

Here we present the scripts and data corresponding with above stated publication. 
We performed several in silico analyses, including methylation quantitative trait loci (*cis*-mQTL) analysis using QTLtool v1.1 and Two Sample Mendelian Randomisation (2SMR) using TwoSampleMR v0.4.22.

Data was obtained from several sources. mQTL analysis was performed in genotyping and methylation data from the AtheroExpress Biobank^1^. 2SMR was performed using CARDIoGRAM+C4D^2^ and METASTROKE^3^ GWAS summary statistics.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Genetics/Artikel Scriptie/DATA/v19 mQTLs")

library(readr)
library(rmarkdown)
library(data.table)
library(tidyverse)
library(kableExtra)
library(docxtools)



```

```{r setwd, include=FALSE}

```

## *cis*-mQTL analysis {.tabset .tabset-fade .tabset-pills}

*cis*-mQTL was perfomed using readily built scripts by S.W. Van der Laan, Associate Professor UMC Utrecht. 

### Data preparation

We extracted our regions of interest from the Illumina 450k manifest file. Only looked at our 11 genes of interest and only at CpGs within -2000 TSS and 1st exon. A range of +/- 250kb was used on this region of interest to analyse for mQTLs using QTLtools^4^.

```{r data preparation, message=FALSE, warning=FALSE}

#read data

CPG <- read.delim("CPG.txt", header = FALSE)

Illumina450K <- read_csv("HumanMethylation450_15017482_v1-2.csv")

#extract genes of interest

IlluminaREGIONS <- Illumina450K[grep("KLF4|HOXA5|TMEM184B|ADAMTSL5|KLF3|CMKLR1|PKP4|ACVRL1|DOK4|SPRY2|ENOSF1",
                                  Illumina450K$UCSC_RefGene_Name), ]

Illumina_data <- IlluminaREGIONS[-grep("Body|3'UTR", IlluminaREGIONS$UCSC_RefGene_Group), ]

ROI <- data.frame(Illumina_data$IlmnID, Illumina_data$UCSC_RefGene_Name, Illumina_data$Chromosome_36, Illumina_data$Coordinate_36, Illumina_data$Coordinate_36, Illumina_data$Strand)

#Range, lead, trait column added

range <- numeric(157)

ROI["range"] <- range

ROI$range <- 250000


lead <- numeric(157)
ROI["lead"] <- lead

ROI$lead <- "new_lead"

trait <- numeric(157)
ROI["trait"] <- trait

ROI$trait <- "SS"

#prepare first row

ROI[1, 8] <- "previous_&_new_lead"

#to and from column added

ROI$from <- ROI$Illumina_data.Coordinate_36

ROI$to <- ROI$Illumina_data.Coordinate_36

#change names

colnames(ROI)[1:4] <- c("CpG", "Gene", "Chr", "Location")

#re-order

ROI2 <- ROI[c(1, 2, 3, 4, 10, 11, 7, 8, 9)]

#from to change values to TSS and -/+ 250kb

REGIONS <- ROI2

REGIONS$from[REGIONS$from %like% "^1590"] <- "158771722"
REGIONS$from[REGIONS$from %like% "^383"] <- "38092185"
REGIONS$from[REGIONS$from %like% "^271"] <- "27149812"
REGIONS$from[REGIONS$from %like% "^109"] <- "109291576"
REGIONS$from[REGIONS$from %like% "^107"] <- "107257216"
REGIONS$from[REGIONS$from %like% "^505"] <- "50342380"
REGIONS$from[REGIONS$from %like% "^798"] <- "79813087"
REGIONS$from[REGIONS$from %like% "^560"] <- "56077886"
REGIONS$from[REGIONS$from %like% "^702"] <- "702676"
REGIONS$from[REGIONS$from %like% "^703"] <- "702676"
REGIONS$from[REGIONS$from %like% "^146"] <- "1464019"
REGIONS$from[REGIONS$from %like% "^369"] <- "36998962"


REGIONS$to[REGIONS$to %like% "^1590"] <- "159021722"
REGIONS$to[REGIONS$to %like% "^383"] <- "38342185"
REGIONS$to[REGIONS$to %like% "^271"] <- "27399812"
REGIONS$to[REGIONS$to %like% "^109"] <- "109541576"
REGIONS$to[REGIONS$to %like% "^107"] <- "107507216"
REGIONS$to[REGIONS$to %like% "^505"] <- "50592380"
REGIONS$to[REGIONS$to %like% "^798"] <- "80063087"
REGIONS$to[REGIONS$to %like% "^560"] <- "56327886"
REGIONS$to[REGIONS$to %like% "^702"] <- "952676"
REGIONS$to[REGIONS$to %like% "^703"] <- "952676"
REGIONS$to[REGIONS$to %like% "^146"] <- "1714019"
REGIONS$to[REGIONS$to %like% "^369"] <- "37248962"

Inputfile <- REGIONS


#export

#write.table(ROI2, file = "ROI", sep = " ", row.names = FALSE, col.names = FALSE)


```

### Input data

```{r, layout="l-body-outset", message=FALSE}

Inputfile %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)

```

### Results {.tabset .tabset-fade .tabset-pills}

#### Nominal results

Fraction of the filtered post QC-results
(*full table available upon request*)

```{r nom results, message=FALSE, warning=FALSE}
temp <- list.files(pattern = "DEFAULT.nominal")

mQTL_data_nom <- lapply(temp, read.csv)

mQTL_data_nom <- do.call(rbind, mQTL_data_nom)

head(mQTL_data_nom) %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>%
    kable(.) %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)

```

#### Permutation results

Filtered post QC-results

```{r perm results, message=FALSE, warning=FALSE}

temp <- list.files(pattern = "perm.Q")

mQTL_data_perm <- lapply(temp, read.csv)

mQTL_data_perm <- do.call(rbind, mQTL_data_perm)

mQTL_data_perm %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>%
    kable(.) %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)
```

## Filter out all non-promoter CpG sites

To assess our hypothesis that promoter methylation has an effect on cardiovascular risk, we had to filter out the non-promoter CpG sites. This was performed since mQTL analysis used al CpG sites within the large +/- 250kb region.

```{r promoter filtering, message=FALSE, warning=FALSE}

# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# 
# BiocManager::install("IlluminaHumanMethylation450kanno.ilmn12.hg19")

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)

#load data
data(IlluminaHumanMethylation450kanno.ilmn12.hg19)

data(Locations)

Illumina_data <- as.data.frame(Locations)

#read the table I made (by hand...) using the UCSC browser Human Feb. 2009 (GRCh37/hg19)
Gene_regions <- read_delim("Gene_regions.csv", 
                           ";", escape_double = FALSE, trim_ws = TRUE)

#extract CpG sites (quick and neat :), not so efficient)

HOXA5 <- subset(Illumina_data, pos > 27182665 & pos < 27185287 & chr == "chr7")
TMEM184B <- subset(Illumina_data, pos > 38668890 & pos < 38671040 & chr == "chr16")
ADAMTSL5 <- subset(Illumina_data, pos > 1512962 & pos < 1515188 & chr == "chr19")
KLF4 <- subset(Illumina_data, pos > 110251449 & pos < 110253927 & chr == "chr9")
KLF3 <- subset(Illumina_data, pos > 38663817 & pos < 38666082 & chr == "chr4")
CMKLR1 <- subset(Illumina_data, pos > 108732804 & pos < 108735094 & chr == "chr12")
PKP4 <- subset(Illumina_data, pos > 159311611 & pos < 159313730 & chr == "chr2")
ACVRL1 <- subset(Illumina_data, pos > 52298202 & pos < 52301479 & chr == "chr12")
DOK4 <- subset(Illumina_data, pos > 57520217 & pos < 57522407 & chr == "chr16")
SPRY2 <- subset(Illumina_data, pos > 80914757 & pos < 80917086 & chr == "chr13")
ENOSF1 <- subset(Illumina_data, pos > 712504 & pos < 715676 & chr == "chr18")

CpGs_mQTLs <- rbind(HOXA5, TMEM184B, ADAMTSL5, KLF4, KLF3, CMKLR1, PKP4, ACVRL1, DOK4, SPRY2, ENOSF1)

#order the data
CpGs_mQTLs$CpG <- rownames(CpGs_mQTLs)
rownames(CpGs_mQTLs) <- c()

CpGs_mQTLs <- CpGs_mQTLs[order(CpGs_mQTLs$chr), ]

CpGs_mQTLs <- CpGs_mQTLs[, c(4, 3, 1, 2)]

#list of CpGs that are in my regions
CpGs_region <- list(unique(CpGs_mQTLs$CpG))

#filter out shear stress mQTLs based on CpG sites

mQTL_data <- mQTL_data_perm[mQTL_data_perm$ProbeID %in% CpGs_region[[1]], ]

#unify RS-identifiers for mQTL data

mQTL_data$VARIANT <- as.character(mQTL_data$VARIANT)

has.rsid = grep(mQTL_data$VARIANT, pattern="rs")
hasnt.rsid = setdiff(1:nrow(mQTL_data), has.rsid)


mQTL_data[has.rsid, "VARIANT"] <-  gsub(x = unlist(mQTL_data[has.rsid, "VARIANT"]), pattern = "(rs\\d+):.*$", replacement = "\\1")
mQTL_data[hasnt.rsid, "VARIANT"] <- gsub(x = unlist(mQTL_data[hasnt.rsid, "VARIANT"]), pattern = "(:[ACTG].*$)", replacement = "")

#table

mQTL_data %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>%
    kable(.) %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)

#write.csv(TEST, file = "mQTLs_shear_stress_nominal_v20_filtered.csv", row.names = F)
```




## Two Sample MR analysis {.tabset .tabset-fade .tabset-pills}

mQTLs passing quality control were used for [2SMR analysis](https://mrcieu.github.io/TwoSampleMR/). These SNPs and SNPs in LD were identified in CARDIoGRAM+C4D and METASTROKE GWAS summary statistics using the TwoSampleMR tool^5^ in R. Causility was determined with the passing SNPs.



### CARDIoGRAM+C4D {.tabset .tabset-fade .tabset-pills}

2SMR analysis of shear stress associated DNA methylation on CHD.

#### 2SMR and horizontal pleiotropy results
```{r DNAm vs CHD, echo=TRUE, message=FALSE, warning=FALSE}
#plaque MR analysis with CHD as outcome
#2SMR analyse

library(TwoSampleMR)
library(MRInstruments)
data("gwas_catalog")
ao <- available_outcomes()

#read data manually

write.csv(mQTL_data, "mQTL_data.csv")

mQTL_exposure_data <- read_exposure_data(
  filename = "mQTL_data.csv",
  sep = ",",
  snp_col = "VARIANT",
  beta_col = "Beta",
  se_col = "SE",
  effect_allele_col = "CodedAlleleA",
  other_allele_col = "OtherAlleleB",
  eaf_col = "MAF",
  pval_col = "Perm_P",
  gene_col = "GeneName",
  samplesize_col = "N"
)

mQTL_exposure_data$exposure <- "DNAm"

##MR on Coronary Artery Disease

#clumping

mQTL_exposure_data_clumped <- clump_data(mQTL_exposure_data)

#extract outcome data

out_dat_7 <- extract_outcome_data(
  snps = mQTL_exposure_data_clumped$SNP,                  
  outcomes = 7,                                           #enter your MR Base outcomes here
  proxies = TRUE
)

#harmonize data

mQTL_vs_out_7 <- harmonise_data(
  exposure_dat = mQTL_exposure_data_clumped,
  outcome_dat = out_dat_7
)


#MR analysis

mQTL_res_out_7 <- mr(mQTL_vs_out_7)

mQTL_res_out_7 %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)

res_single_out_7 <- mr_singlesnp(mQTL_vs_out_7)

res_single_out_7 %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)

res_pleiotropy_7 <- mr_pleiotropy_test(mQTL_vs_out_7)

res_pleiotropy_7 %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)
```

#### Scatter plot and forest plot with passing SNPs

```{r plot CHD, message=FALSE, warning=FALSE}
library(ggplot2)

PLOT_7 <- mr_scatter_plot(mQTL_res_out_7, mQTL_vs_out_7)

#to make the plot my own the identifier needs to be named value

plotlot <- mQTL_res_out_7[1,1]
plotlot. <- ".7"

plottype <- paste0(plotlot, plotlot.)

PLOT_7 <- PLOT_7[[plottype]] + #COPY THE ID FROM THE MR TABLE!
  theme_minimal() + 
  theme(legend.position="bottom",
        panel.background = element_rect(fill = "#FFFFFF",
                                        colour = "#FFFFFF"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "black")) +
  labs(x = "SNP effect on DNA methylation") + 
  labs(y = "SNP effect on Coronary Heart Disease") +
  scale_color_manual(values=c("#5EB17F","#F59D10","#E55738","#8D5B9A","#E35493"))

PLOT_7

#forestplot

PLOT_7_forest <- mr_forest_plot(res_single_out_7)
PLOT_7_forest[[1]]

```


### METASTROKE {.tabset .tabset-fade .tabset-pills}

2SMR analysis of shear stress associated DNA methylation on Stroke

#### 2SMR and horizontal pleiotropy results
```{r DNAm vs IS, echo=TRUE, message=FALSE, warning=FALSE}
#plaque MR analysis with IS as outcome
#2SMR analyse

library(TwoSampleMR)
library(MRInstruments)
data("gwas_catalog")
ao <- available_outcomes()

#read data manually

write.csv(mQTL_data, "mQTL_data.csv")

mQTL_exposure_data <- read_exposure_data(
  filename = "mQTL_data.csv",
  sep = ",",
  snp_col = "VARIANT",
  beta_col = "Beta",
  se_col = "SE",
  effect_allele_col = "CodedAlleleA",
  other_allele_col = "OtherAlleleB",
  eaf_col = "MAF",
  pval_col = "Perm_P",
  gene_col = "GeneName",
  samplesize_col = "N"
)

mQTL_exposure_data$exposure <- "DNAm"

##MR on Coronary Artery Disease

#clumping

mQTL_exposure_data_clumped <- clump_data(mQTL_exposure_data)

#extract outcome data

out_dat_1108 <- extract_outcome_data(
  snps = mQTL_exposure_data_clumped$SNP,                  
  outcomes = 1108,                                           #enter your MR Base outcomes here
  proxies = TRUE
)

#harmonize data

mQTL_vs_out_1108 <- harmonise_data(
  exposure_dat = mQTL_exposure_data_clumped,
  outcome_dat = out_dat_1108
)


#MR analysis

mQTL_res_out_1108 <- mr(mQTL_vs_out_1108)

mQTL_res_out_1108 %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)

res_single_out_1108 <- mr_singlesnp(mQTL_vs_out_1108)

res_single_out_1108 %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)

# res_pleiotropy_1108 <- mr_pleiotropy_test(mQTL_vs_out_1108)
# 
# res_pleiotropy_1108 %>%
#   kable() %>%
#   kable_styling(bootstrap_options = c("striped", "condensed"), font_size = 12, fixed_thead = T)
```

#### Scatter plot and forest plot with passing SNPs

Not enough SNPs for a plot.


## Regional Association Plot ENOSF1

We made the regional association plot of the ENOSF1 region using Regional Association ComparER (RACER) v0.1 of [oliviasabik](https://github.com/oliviasabik/RACER).

```{r RAP, echo=FALSE, message=FALSE, warning=FALSE}
# install.packages("devtools") 
# library(devtools) 
# install_github("oliviasabik/RACER") 

library(RACER)


#read data
mQTLs_shear_stress_v20 <- read_csv("aegs_QC_qtlnom_region_11_excl_DEFAULT.nominal.all.txt")

mQTLs_shear_stress_v20[,-1] <- gsub("\\:.*", "", as.matrix(mQTLs_shear_stress_v20[,-1]))


#correct format for RACER
mQTLs_shear_stress_v20$Nominal_P <- as.numeric(gsub(",", ".", mQTLs_shear_stress_v20$Nominal_P))

mQTLs_shear_stress_nominal_v20 <- mQTLs_shear_stress_v20[order(mQTLs_shear_stress_v20$Nominal_P), ]

RAP <- mQTLs_shear_stress_nominal_v20[!duplicated(mQTLs_shear_stress_nominal_v20$VARIANT), ]



#Regional association plot

mQTLs_plot_new <- RACER::formatRACER(assoc_data = RAP, 
                                     chr_col = 3, 
                                     pos_col = 4, 
                                     p_col = 32)

#Add LD data

mQTLs_plot_new_2 <- RACER::ldRACER(assoc_data = mQTLs_plot_new, 
                                   rs_col = 2, pops = "EUR", 
                                   lead_snp = "rs7235957") 

#edit the function to add some features-----

rasplot <- function (assoc_data, chr, build = "hg19", set = "protein_coding", 
          plotby, gene_plot = NULL, snp_plot = NULL, start_plot = NULL, 
          end_plot = NULL, label_lead = FALSE) 
{
  if (missing(assoc_data)) {
    stop("Please provide a data set to plot.")
  }
  else if (missing(chr)) {
    stop("Please specify which chromosome you wish to plot.")
  }
  else if (missing(plotby)) {
    stop("Please specify the method by which you wish to plot.")
  }
  else if (plotby == "gene") {
    if (is.null(gene_plot)) {
      stop("Please specify a gene to plot by.")
    }
  }
  else if (plotby == "snp") {
    if (is.null(snp_plot)) {
      stop("Please specify a snp to plot by.")
    }
  }
  else if (plotby == "coord") {
    if (is.null(start_plot) | is.null(end_plot)) {
      stop("Please specify start coordinate for plot.")
    }
  }
  else {
    message("All inputs are go.")
  }
  reqs = c("CHR", "POS", "LOG10P")
  cols = colnames(assoc_data)
  if (sum(reqs %in% cols) == 3) {
  }
  else {
    stop("Association Data Set is missing a required column, please format your data set using formatRACER.R.")
  }
  reqs_2 = c("LD", "LD_BIN")
  if (sum(reqs_2 %in% cols) == 2) {
  }
  else {
    message("Association Data Set is missing LD data, the resulting plot won't have LD information, but you can add it using the ldRACER.R function.")
  }
  `%>%` <- magrittr::`%>%`
  if (build == "hg38") {
    utils::data(hg38)
    chr_in = chr
    colnames(hg38) = c("GENE_ID", "CHR", "TRX_START", "TRX_END", 
                       "LENGTH", "GENE_NAME", "TYPE")
    gene_sub = hg38[hg38$CHR == chr_in, ]
  }
  else if (build == "hg19") {
    utils::data(hg19)
    chr_in = chr
    colnames(hg19) = c("GENE_ID", "CHR", "TRX_START", "TRX_END", 
                       "LENGTH", "GENE_NAME", "TYPE")
    gene_sub = hg19[hg19$CHR == chr_in, ]
  }
  if (set == "protein_coding") {
    gene_sub = gene_sub[gene_sub$TYPE == "protein_coding", 
                        ]
  }
  else {
    gene_sub = gene_sub
  }
  if (sum(is.null(plotby)) == 1) {
    stop("Please specify a method by which to plot.")
  }
  if (sum(is.null(plotby)) == 0) {
    message("Plotting by...")
    if ((plotby == "coord") == TRUE) {
      message("coord")
      start = start_plot
      end = end_plot
    }
    else if ((plotby == "gene") == TRUE) {
      message(paste("gene:", gene_plot))
      if (sum(is.null(gene_plot)) == 0) {
        p = subset(gene_sub, gene_sub$GENE_NAME == gene_plot)
        start = min(p$TRX_START) - 5e+05
        end = max(p$TRX_END) + 5e+05
      }
      else {
        message("No gene specified.")
      }
    }
    else if ((plotby == "snp") == TRUE) {
      message(paste("snp", snp_plot))
      q = assoc_data[assoc_data$RS_ID == snp_plot, ]
      w = q$POS
      w = as.numeric(as.character(w))
      start = w - 5e+05
      end = w + 5e+05
    }
  }
  gene_sub = subset(gene_sub, gene_sub$TRX_START > (start - 
                                                      50000))
  gene_sub = subset(gene_sub, gene_sub$TRX_END < (end + 50000))
  gene_sub = gene_sub[, c(3, 4, 6)]
  gene_sub = reshape2::melt(gene_sub, id.vars = "GENE_NAME")
  gene_sub$y_value = as.numeric(as.factor(gene_sub$GENE_NAME))
  plot_lab = subset(gene_sub, gene_sub$variable == "TRX_END")
  message("Reading in association data")
  in.dt <- as.data.frame(assoc_data)
  in.dt$POS = as.numeric(as.character(in.dt$POS))
  in.dt$LOG10P = as.numeric(as.character(in.dt$LOG10P))
  in.dt$CHR = as.numeric(as.character(in.dt$CHR))
  in.dt = dplyr::filter_(in.dt, ~CHR == chr_in)
  in.dt = dplyr::filter_(in.dt, ~POS > start) %>% dplyr::filter_(~POS < 
                                                                   end)
  if (label_lead == TRUE) {
    lsnp_row = which(in.dt$LABEL == "LEAD")
    label_data = in.dt[lsnp_row, ]
    if (dim(label_data)[1] == 0) {
      lsnp_row = in.dt[in.dt$LOG10P == max(in.dt$LOG10P), 
                       ]
      label_data = lsnp_row[1, ]
    }
  }
  message("Generating Plot")
  
  uithof_color = c("#FBB820","#F59D10","#E55738","#DB003F","#E35493","#D5267B",
                   "#CC0071","#A8448A","#9A3480","#8D5B9A","#705296","#686AA9",
                   "#6173AD","#4C81BF","#2F8BC9","#1290D9","#1396D8","#15A6C1",
                   "#5EB17F","#86B833","#C5D220","#9FC228","#78B113","#49A01D",
                   "#595A5C","#A2A3A4", "#D7D8D7", "#ECECEC", "#FFFFFF", "#000000")
  
  if ("LD" %in% colnames(in.dt) && "LD_BIN" %in% colnames(in.dt)) {
    c = ggplot2::ggplot(gene_sub, ggplot2::aes_string(x = "value", 
                                                      y = "y_value")) + 
      geom_line(aes_string(group = "GENE_NAME"), 
                size = 2.5, 
                color = "black",
                alpha = 0.8) +
      geom_line(ggplot2::aes_string(group = "GENE_NAME"), 
                         size = 2, 
                         color = uithof_color[23],
                         alpha = 0.6) +
      ggplot2::theme_minimal() + ggplot2::geom_text(data = plot_lab, 
                                               ggplot2::aes_string(x = "value", y = "y_value", label = "GENE_NAME"), 
                                               hjust = -0.1, vjust = 0.3, size = 2) + 
      ggplot2::xlab(paste0("Chromosome ", chr_in, " Position")) + 
      ggplot2::coord_cartesian(xlim = c(start, end), ylim = c(0, (max(gene_sub$y_value) + 1))) +
      ggplot2::theme(axis.title.y = ggplot2::element_text(color = "white", size = 26, vjust = 1, hjust = 1),
                     axis.title.x = element_text(size = 10),
                     axis.text.y = ggplot2::element_blank(), 
                     axis.ticks.y = ggplot2::element_blank(),
                     panel.grid.major.y = element_blank(),
                     panel.grid.minor = element_blank(),
                     panel.grid.major.x = element_line(colour = "#DEDEDE", size = 0.5, linetype = "longdash", lineend = "round"),
                     axis.line.x = element_line(colour = uithof_color[30]),
                     axis.text.x = element_text(colour = uithof_color[30], size = 6),
                     plot.margin=unit(c(0.2,0,0.2,0.2), "cm"))
    
    
    
    b = ggplot2::ggplot(in.dt, ggplot2::aes_string(x = "POS", 
                                                   y = "LOG10P", color = "LD_BIN")) + ggplot2::geom_point(size = 2) + 
      ggplot2::scale_colour_manual(expression(paste(R^2, sep = "")), values = c(`1.0-0.8` = uithof_color[11], 
                                              `0.8-0.6` = uithof_color[2], `0.6-0.4` = uithof_color[8], 
                                              `0.4-0.2` = uithof_color[24], `0.2-0.0` = uithof_color[18],
                                              `NA` = "grey"), drop = FALSE) + ggplot2::theme_minimal() +
      ggplot2::theme(legend.title = ggplot2::element_text(size = 10),
                     legend.text = element_text(size = 6),
                     legend.title.align = 0.15,
                     panel.grid.major.x = element_blank(),
                     panel.grid.minor = element_blank(),
                     panel.grid.major.y = element_line(colour = "#DEDEDE", size = 0.5, linetype = "longdash", lineend = "round"),
                     axis.line = element_line(colour = uithof_color[30]),
                     axis.text.x = element_blank(), #axis.text.x = element_text(colour = uithof_color[30], size = 6)
                     axis.text.y = element_text(colour = uithof_color[30], size = 6),
                     axis.title.y = element_text(size = 10),
                     axis.title.x = element_blank(),
                     plot.margin=unit(c(0.2,0,-0.2,0.2), "cm")) + #add more stuff!!!
      ggplot2::xlab("Chromosome Position") + ggplot2::ylab("-log10(p-value)") + 
      ggplot2::coord_cartesian(xlim = c(start, end), ylim = c(min(in.dt$LOG10P), 
                                                              max(in.dt$LOG10P)))
  }
  else {
    c = ggplot2::ggplot(gene_sub, ggplot2::aes_string(x = "value", 
                                                      y = "y_value")) + ggplot2::geom_line(ggplot2::aes_string(group = "GENE_NAME"), 
                                                                                           size = 2) + ggplot2::theme_bw() + ggplot2::geom_text(data = plot_lab, 
                                                                                                                                                ggplot2::aes_string(x = "value", y = "y_value", label = "GENE_NAME"), 
                                                                                                                                                hjust = -0.1, vjust = 0.3, size = 2.5) + ggplot2::theme(axis.title.y = ggplot2::element_text(color = "white", 
                                                                                                                                                                                                                                             size = 28), axis.text.y = ggplot2::element_blank(), 
                                                                                                                                                                                                        axis.ticks.y = ggplot2::element_blank()) + ggplot2::xlab(paste0("Chromosome ", 
                                                                                                                                                                                                                                                                        chr_in, " Position")) + ggplot2::coord_cartesian(xlim = c(start, 
                                                                                                                                                                                                                                                                                                                                  end), ylim = c(0, (max(gene_sub$y_value) + 1)))
    b = ggplot2::ggplot(in.dt, ggplot2::aes_string(x = "POS", 
                                                   y = "LOG10P")) + ggplot2::geom_point() + ggplot2::theme_bw() + 
      ggplot2::xlab("Chromosome Position") + ggplot2::ylab("-log10(p-value)") + 
      ggplot2::coord_cartesian(xlim = c(start, end), ylim = c(min(in.dt$LOG10P), 
                                                              max(in.dt$LOG10P)))
  }
  if (label_lead == TRUE) {
    b = b + ggplot2::geom_point(data = label_data, aes_string(x = "POS", 
                                                     y = "LOG10P"), color = "black", size = 2.5)
    b = b + ggplot2::geom_text(data = label_data, aes_string(label = "RS_ID"), 
                      color = "black", size = 3, hjust = 1.05, nudge_y = 1.5, nudge_x = 2)
  }
  ggpubr::ggarrange(b, c, heights = c(3, 1), nrow = 2, ncol = 1, 
                    common.legend = TRUE, legend = "right")
}


#plot the data----

ENOSF1_region_plot <- rasplot(assoc_data = mQTLs_plot_new_2, 
                       chr = 18, 
                       build = "hg19", 
                       plotby = "coord", 
                       start_plot = 463009, 
                       end_plot = 965639,
                       label_lead = T)


ENOSF1_region_plot


```

## Data availability {.tabset .tabset-fade .tabset-pills}

### CpG list

```{r echo=FALSE}

embed_data= function(x= mtcars, filename= "file.csv", label= "CpG list"){

  # Create encoded Base64 datastream 
  encode_data= function(x){
    write.csv2(x, "./file.csv")
    enc= sprintf('data:text/csv;base64,%s', openssl::base64_encode(paste0(readLines("./file.csv"), collapse="\n")) )
    unlink("./file.csv")
    return(enc)
  }

  # String result ready to be placed in rmarkdown
  paste0("<a download='", filename, "' href=", encode_data(x), ">", label, "</a>")

}
```

`r embed_data(CPG, filename="cpg.csv")` 

### Nominal mQTL results

```{r echo=FALSE}

embed_data= function(x= mtcars, filename= "file.csv", label= "Nominal mQTL results"){

  # Create encoded Base64 datastream 
  encode_data= function(x){
    write.csv2(x, "./file.csv")
    enc= sprintf('data:text/csv;base64,%s', openssl::base64_encode(paste0(readLines("./file.csv"), collapse="\n")) )
    unlink("./file.csv")
    return(enc)
  }

  # String result ready to be placed in rmarkdown
  paste0("<a download='", filename, "' href=", encode_data(x), ">", label, "</a>")

}

embed_data= function(x= mtcars, filename= "file.csv", label= "Only region 11 for association plot"){

  # Create encoded Base64 datastream 
  encode_data= function(x){
    write.csv2(x, "./file.csv")
    enc= sprintf('data:text/csv;base64,%s', openssl::base64_encode(paste0(readLines("./file.csv"), collapse="\n")) )
    unlink("./file.csv")
    return(enc)
  }

  # String result ready to be placed in rmarkdown
  paste0("<a download='", filename, "' href=", encode_data(x), ">", label, "</a>")

}
```

Available upon request

### Permutation mQTL results

```{r echo=FALSE}

embed_data2= function(x= mtcars, filename= "file.csv", label= "Permutation mQTL results"){

  # Create encoded Base64 datastream 
  encode_data= function(x){
    write.csv2(x, "./file.csv")
    enc= sprintf('data:text/csv;base64,%s', openssl::base64_encode(paste0(readLines("./file.csv"), collapse="\n")) )
    unlink("./file.csv")
    return(enc)
  }

  # String result ready to be placed in rmarkdown
  paste0("<a download='", filename, "' href=", encode_data(x), ">", label, "</a>")

}
```

`r embed_data2(mQTL_data_perm, filename="mQTL_data_perm.csv")`

### Gene regions for promoter filtering

```{r echo=FALSE}

embed_data3= function(x= mtcars, filename= "file.csv", label= "Gene regions for filtering"){

  # Create encoded Base64 datastream 
  encode_data= function(x){
    write.csv2(x, "./file.csv")
    enc= sprintf('data:text/csv;base64,%s', openssl::base64_encode(paste0(readLines("./file.csv"), collapse="\n")) )
    unlink("./file.csv")
    return(enc)
  }

  # String result ready to be placed in rmarkdown
  paste0("<a download='", filename, "' href=", encode_data(x), ">", label, "</a>")

}


# Commit - "upload all scripts"
```

`r embed_data3(Gene_regions, filename="Gene_regions.csv")` 





## References

1. Verhoeven, B. A. N. et al. Athero-express: differential atherosclerotic plaque expression of mRNA and protein in relation to cardiovascular events and patient characteristics. Rationale and design. Eur. J. Epidemiol. 19, 1127–1133 (2004)
2. Nikpay, M. et al. A comprehensive 1,000 Genomes-based genome-wide association meta-analysis of coronary artery disease. Nat. Genet. 47, 1121–1130 (2015)
3. Malik, R. et al. Low-frequency and common genetic variation in ischemic stroke: The METASTROKE collaboration. Neurology 86, 1217–1226 (2016)
4. Delaneau, O. et al. A complete tool set for molecular QTL discovery and analysis. Nat Commun 8, 15452 (2017)
5. Davey Smith, G. & Hemani, G. Mendelian randomization: genetic anchors for causal inference in epidemiological studies. Hum. Mol. Genet. 23, R89-98 (2014)



------
<sup>&copy; 1998-2019 Ruben Methorst | r.methorst[at]umcutrecht.nl. </sup>
------
